### ISSUE TITLE
please clarify whether parse_sexp should parse atoms

### ISSUE THREAD

**Issue by bremner** [mjsottile/sfsexp#9]
Playing with `tests/ctorture.c` with a file containing "foo", it seems like first parse (when using `ctorture -t`) with `read_one_sexp` succeeds, but the subsequent ones using `parse_sexp` fail, setting `sexp_errno` to  `SEXP_ERR_INCOMPLETE`.
Is this the intended behaviour?  If so I guess I can make my own more permissive wrapper, but it seems like it might be a generally useful behaviour, i.e. `read` in most lisps is fine with reading atoms.

**Comment by mjsottile** [comment]
That sounds like an unintended behavior.  I should have time to try to reproduce the behavior and look at what’s going on when some day job tasks clear up soon.

**Comment by mjsottile** [comment]
Ok - I think I understand the issue better.  I believe that what you've run into is a consequence of the way the parser is designed around continuations.  This is due to the library being designed to handle wire protocols that use s-expressions for their format where parsing should be started as data arrives instead of assuming one can read the entire expression into a buffer and then start parsing.  The continuation mechanism is present to allow this to happen for many streams simultaneously without presuming any mechanisms (like thread-local storage) are available.  The parser is intended to work on unbounded streams containing a sequence of s-expressions like:

```
(foo a b c)(foo2 d e f)(bar g h i)
```

The goal was to be able to pop expressions off the stream, retaining the state of the parser representing "the rest of the stream" such that the parser can pick up and pop the next expression off the stream when it is demanded.  The problem is, the stream source may not deliver expressions in chunks corresponding to the expression boundaries : for example, you may see something like:

```
+-------------------+-------+--------+
|(foo a b c)(foo2 d |e f)(ba|r g h i)|
+-------------------+-------+--------+
```

where we observe three chunks containing three distinct expressions, but the chunk boundaries aren't aligned with the expression boundaries.  In the first chunk, the first expression "(foo a b c)" can be parsed and returned as a well formed sexp_t, and the remaining "(foo2 d " is waiting in the continuation to be parsed the next time around.  When the parser hits the end of the partial expression, a continuation is returned indicating that the expression is incomplete.  When the second chunk "e f)(ba" is received, the second expression (foo2 d e f)" is completed and returned, and the continuation contains the partial "(ba", waiting for the last chunk with the "r g h i)" to complete "(bar g h i)".

The odd behavior you observed is a consequence of the input likely containing whitespace that cleanly terminates the atom when the first parse call occurs.  When you pass a single atom in from a file where the buffer looks like "foo\n\0", then the atom "foo" is parsed correctly.  The ctorture example then pretty-prints the foo to a buffer which will contain "foo\0".  The key is that the carriage return that terminates the atom in the input file is not present in the buffer that was printed into.  The parser hits the end of the buffer having seen "foo", but doesn't know if "foo" is the whole atom or if additional buffer will be provided for a larger atom like "food".  As such, it declares that the expression is incomplete since the continuation-based parser doesn't know that nothing else will ever show up.

I sort of remember seeing this issue many years ago, and decided not to worry too much about it since bare atoms were a rare case in the context where the library was developed.

One solution to this would be to force a whitespace character to be present in the ctorture case at the end of the buffer that forces the parser to determine that the buffer represents a complete expression element.  Another option would be to have a special wrapper around the parser that assumes that there will never be additional data provided for the continuation such that any in-progress atoms are treated to have an implicit whitespace character following them to force them to complete.  That would likely be a cleaner approach.

**Comment by mjsottile** [comment]
I just realized while re-reading the code that the “special wrapper” is parse_sexp since that wraps the continuation based parser. A tweak there could address the issue for bare atoms.

**Comment by mjsottile** [comment]
This should be fixed in commit e68e3b7.  parse_sexp should now behave as expected with a bare atom.
